File: frontend/types/niche.ts
import { KeywordOpportunity } from './opportunity';
import { KeywordCluster } from './cluster';

export type MarketSize = 'small' | 'medium' | 'large' | 'huge';
export type CompetitionLevel = 'low' | 'medium' | 'high' | 'very_high';

export interface ContentGap {
  gap_type: string;
  description: string;
  keywords: string[];
  estimated_impact: string;
  priority: string;
}

export interface MarketOpportunity {
  cluster_name: string;
  cluster_theme: string;
  opportunity_score: number;
  opportunity_level: string;
  total_search_volume: number;
  avg_difficulty: number;
  recommended_action: string;
}

export interface NicheAnalysis {
  seed_keyword: string;
  total_keywords: number;
  total_search_volume: number;
  market_size: MarketSize;
  competition_level: CompetitionLevel;
  avg_keyword_difficulty: number;
  monetization_potential: number;
  top_serp_features: string[];
  content_gaps: ContentGap[];
  recommended_strategy: string;
  confidence_score: number;
  opportunities: MarketOpportunity[];
  cluster_count: number;
}

export interface NicheDiscoveryResponse {
  seed_keyword: string;
  opportunities: KeywordOpportunity[];
  clusters: KeywordCluster[];
  niche_analysis: NicheAnalysis;
  summary: {
    total_keywords: number;
    total_clusters: number;
    market_size: MarketSize;
    competition_level: CompetitionLevel;
    confidence_score: number;
  };
}

export interface NicheDiscoveryRequest {
  seed_keyword: string;
  filters?: {
    min_volume?: number;
    max_difficulty?: number;
    min_cpc?: number;
    max_cpc?: number;
    intents?: string[];
  };
  limit?: number;
  include_trends?: boolean;
}

File: backend/app/models/niche.py
"""
Niche Analysis Models
Pydantic models for market and niche analysis
"""
from pydantic import BaseModel, Field
from typing import List, Optional
from enum import Enum


class MarketSize(str, Enum):
    """Market size classification"""
    SMALL = "small"
    MEDIUM = "medium"
    LARGE = "large"
    HUGE = "huge"


class CompetitionLevel(str, Enum):
    """Competition level classification"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    VERY_HIGH = "very_high"


class ContentGap(BaseModel):
    """Identified content opportunity gap"""
    gap_type: str = Field(..., description="Type of content gap")
    description: str = Field(..., description="Gap description")
    keywords: List[str] = Field(..., description="Related keywords")
    estimated_impact: str = Field(..., description="Potential impact (low/medium/high)")
    priority: str = Field(..., description="Priority level")


class MarketOpportunity(BaseModel):
    """Specific market opportunity"""
    cluster_name: str
    cluster_theme: str
    opportunity_score: float
    opportunity_level: str
    total_search_volume: int
    avg_difficulty: float
    recommended_action: str


class NicheAnalysis(BaseModel):
    """Complete niche market analysis"""
    seed_keyword: str
    total_keywords: int
    total_search_volume: int
    market_size: MarketSize
    competition_level: CompetitionLevel
    avg_keyword_difficulty: float
    monetization_potential: float
    top_serp_features: List[str]
    content_gaps: List[ContentGap]
    recommended_strategy: str
    confidence_score: float
    opportunities: List[MarketOpportunity]
    cluster_count: int

    class Config:
        use_enum_values = True
